src/core/algorithms/gwo.h

```diff
@@ -0,0 +1,130 @@
+* Neither the name of the <organization> nor the
+names of its contributors may be used to endorse or promote products
+derived from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#ifndef DNN_OPT_CORE_ALGORITHMS_GWO
+#define DNN_OPT_CORE_ALGORITHMS_GWO
+
+#include <core/base/algorithm.h>
+#include <core/generators/uniform.h>
+#include <core/generators/constant.h>
+
+namespace dnn_opt
+{
+namespace core
+{
+namespace algorithms
+{
+
+/**
+ * @brief The GWO class implements an optimization metaheuristic algorithm
+ * called Gray Wolf Optimization (GWO). This is a population based algorithm
+ * inspired in the hunting behavior of gray wolves.
+ *
+ * @author Randy Alonso Benitez <rbenitez@estudiantes.uci.cu>
+ * @version 1.0
+ * @date November, 2019
+ */
+class gwo : public virtual algorithm
+{
+public:
+
+  template<class t_solution>
+  static gwo* make(set<t_solution>* solutions);
+
+  /**
+   * @copydoc algorithm::reset()
+   *
+   * Re-generate the solutions, set hyper-parameters to its defaults values.
+   */
+  virtual void reset() override;
+
+  virtual void optimize() override;
+
+  using algorithm::optimize;
+
+  /**
+   * @brief Update the values ​​of the three best agents.
+   */
+  virtual void update_elite() override;
+
+  virtual solution* get_best() override;
+
+  virtual void set_params(std::vector<float> &params) override;
+
+  using algorithm::set_params;
+
+  /**
+   * @brief The basic destructor of this class.
+   */
+  virtual ~gwo() override;
+
+protected:
+
+  virtual void init() override;
+  /**
+   * @brief The basic contructor of a gwo class.
+   *
+   * @param solutions a set of individuals.
+   */
+  template<class t_solution>
+  gwo(set<t_solution>* solutions);
+
+  solution* _alpha;
+  solution* _beta ;
+  solution* _delta;
+  float* _a ;
+  float* _r1;
+  float* _r2;
+  float* _A1;
+  float* _C1;
+  float* _r;
+
+
+};
+
+/* templated function implementations */
+
+template<class t_solution>
+gwo* gwo::make(set<t_solution> *solutions)
+{
+  auto* result = new gwo(solutions);
+  result->init();
+  return result;
+}
+
+template<class t_solution>
+gwo::gwo(set<t_solution>* solutions)
+: algorithm(solutions)
+{
+
```

Comment:
I think we should include a check here to ensure that `solutions->size() > 2` as we need at least three wolves for the alpha, the beta and the delta.